Sources/PromiseKit/Future.swift

```diff
@@ -76,3 +88,21 @@ private extension Future {
     }
   }
 }
+
+extension NSLock {
+  @inlinable
+  @inline(__always)
+  func read<T>(_ work: () -> T) -> T {
+    lock()
+    defer { unlock() }
+    return work()
+  }
+  
+  @inlinable
+  @inline(__always)
+  func write(_ work: () -> Void) {
+    lock()
+    defer { unlock() }
+    return work()
+  }
+}
```

Comment:
Can we move this under PromiseKit/Extensions folder instead?

Reply:
It's not part of the public API, and I think it should not be.

Reply:
What do you mean? It is already, ust put it in a separate file 🤔

Reply:
It is not. For it it be public it has to have a `public modifier`.

Reply:
But if it's internal access, other methods can still see it without exposing the interface right?

Reply:
Decided to mark the extension as `fileprivate` to prevent others using it.